plugins/woocommerce/src/Blocks/InteractivityComponents/Dropdown.php

```diff
@@ -18,60 +18,96 @@ public static function render( $props ) {
 					<div class="components-form-token-field" tabindex="-1">
-						<div class="components-form-token-field__input-container" 
-									data-wc-class--is-active="context.isOpen" 
-									tabindex="-1" 
-									data-wc-on--click="actions.toggleIsOpen" 
-							>
-							<input id="components-form-token-input-1" type="text" autocomplete="off" data-wc-bind--placeholder="state.placeholderText" class="components-form-token-field__input" role="combobox" aria-expanded="false" aria-autocomplete="list" aria-describedby="components-form-token-suggestions-howto-1" value="">
-							<ul hidden data-wc-bind--hidden="!context.isOpen" class="components-form-token-field__suggestions-list" id="components-form-token-suggestions-1" role="listbox">
+						<div class="components-form-token-field__input-container"
+							data-wc-class--is-active="context.isOpen"
+							tabindex="-1"
+							data-wc-on--click="actions.toggleIsOpen"
+						>
+							<?php if ( 'multiple' === $select_type ) { ?>
+								<template
+									data-wc-each="context.selectedItems"
+									data-wc-each-key="context.item.value"
+								>
+									<span class="components-form-token-field__token">
+										<span
+											class="components-form-token-field__token-text"
+											data-wc-text="context.item.label"
+										></span>
+										<button
+											type="button"
+											data-wc-on--click="actions.unselectDropdownItem"
+											data-wc-on--click--parent-action="<?php echo esc_attr( $action ); ?>"
+											class="components-button components-form-token-field__remove-token has-icon"
+										>
+											<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" focusable="false">
+												<path d="M12 13.06l3.712 3.713 1.061-1.06L13.061 12l3.712-3.712-1.06-1.06L12 10.938 8.288 7.227l-1.061 1.06L10.939 12l-3.712 3.712 1.06 1.061L12 13.061z"></path>
+											</svg>
+										</button>
+									</span>
+								</template>
+								<?php foreach ( $selected_items as $selected ) { ?>
+									<span
+										class="components-form-token-field__token"
+										data-wc-key="<?php echo esc_attr( $selected['label'] ); ?>"
+										data-wc-each-child
+									>
+										<span class="components-form-token-field__token-text">
+											<?php echo esc_html( $selected['label'] ); ?>
+										</span>
+										<button
+											type="button"
+											class="components-button components-form-token-field__remove-token has-icon"
+										>
+											<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" focusable="false">
+												<path d="M12 13.06l3.712 3.713 1.061-1.06L13.061 12l3.712-3.712-1.06-1.06L12 10.938 8.288 7.227l-1.061 1.06L10.939 12l-3.712 3.712 1.06 1.061L12 13.061z"></path>
+											</svg>
+										</button>
+									</span>
+								<?php } ?>
```

Comment:
Why do we need this foreach here while we have the `template` above?

At a glance, I think it's for initial render/page load, right?

If yes, and let's assume that we need this foreach. Then I make me question about the benefit of having `template` above. This code can be rewritten like the following to make it work without `wp-each`:

<details>
<summary>Details</summary>

```
							<?php if ( 'multiple' === $select_type ) { ?>
								<?php foreach ( $selected_items as $selected ) { ?>
									<span
										class="components-form-token-field__token"
										data-wc-key="<?php echo esc_attr( $selected['label'] ); ?>"
									>
										<span class="components-form-token-field__token-text">
											<?php echo esc_html( $selected['label'] ); ?>
										</span>
										<button
											type="button"
											data-wc-on--click="actions.unselectDropdownItem"
											data-wc-on--click--parent-action="<?php echo esc_attr( $action ); ?>"
											data-wc-context="<?php echo esc_attr(wp_json_encode( array( 'item' => $selected, 'selectedItems' => $selected_items ) )); ?>"
											class="components-button components-form-token-field__remove-token has-icon"
										>
											<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" focusable="false">
												<path d="M12 13.06l3.712 3.713 1.061-1.06L13.061 12l3.712-3.712-1.06-1.06L12 10.938 8.288 7.227l-1.061 1.06L10.939 12l-3.712 3.712 1.06 1.061L12 13.061z"></path>
											</svg>
										</button>
									</span>
								<?php } ?>
							<?php } ?>

```

</details>

Can you explain what I'm missing here? I'm not against adding `wp-each`, but in this case, I don't see a clear benefit of using it.

Reply:
@dinhtungdu yes, while we don't have the SSR support for each in WC, we must render the each times ourselves too. I agree its a bit more generated markup and thats less ideal, but for the selected items the benefit is being able to update them in f/e context easily while still allowing them to contain nested HTML content which we need for these dropdown selected items.

This approach means the consumer of dropdown doesn't have to perform a navigate to update selected items which was the whole problem in the first place. This Dropdown should not *need* it's consumer to perform navigate to work and each effortlessly solves this.

Is it 100% necessary right this minute? maybe not since we use the filter params to navigate right now as a result of the dropdown change, but this allows us to do client side updates in a very simple way and for selected items that felt perfect to me. This also gives an example for others to follow as I expect each directive to become increasingly useful as we build more reusable components in future.

Reply:
Thanks for your reply! Yes it'd would be way better when we have SSR support and I think this is a good preparation as well as example. Moreover, as you mentioned in our dm, the new `wp-each` will enhance the UX dramatically for shopper with slow network, I think that benefit outperforms the cost of having extra foreach! So I'm happy to mark this as resolved.

> so the other thing is, we can have dropdown be optimistic again
> if there is a slow network connection navigate can take significant time to update state of the dropdown